clients/vault/src/oracle/types.rs

```diff
@@ -22,47 +26,103 @@ pub(crate) type SlotEncodedMap = BTreeMap<Slot, SerializedData>;
 /// Sometimes not enough `StellarMessage::ScpMessage(...)` are sent per slot;
 /// or that the `Stellar:message::TxSet(...)` took too long to arrive (may not even arrive at all)
 /// So I've kept both of them separate: the `EnvelopesMap` and the `TxSetMap`
-pub(crate) type EnvelopesMap = VecDeque<(Slot, Vec<ScpEnvelope>)>;
-pub(crate) type TxSetMap = VecDeque<(Slot, TransactionSet)>;
+pub(crate) type EnvelopesMap = LimitedFifoMap<Slot, Vec<ScpEnvelope>>;
+pub(crate) type TxSetMap = LimitedFifoMap<Slot, TransactionSet>;
 
 pub(crate) type SlotList = BTreeMap<Slot, ()>;
 
-/// This is used to add some helpers that make using the VecDeque easier.
-pub trait LifoMap<T, K> {
-	fn get_with_key(&self, key: &K) -> Option<&T>;
-
-	fn set_with_key(&mut self, key: K, value: T);
-
-	fn remove_with_key(&mut self, key: &K) -> Option<T>;
-
-	fn contains_key(&self, key: &K) -> bool;
+#[derive(Debug, Clone)]
+pub struct LimitedFifoMap<K, T> {
+	limit: usize,
+	queue: VecDeque<(K, T)>,
 }
 
-impl<T, K> LifoMap<T, K> for VecDeque<(K, T)>
+impl<K, T> LimitedFifoMap<K, T>
 where
-	K: PartialEq,
+	K: Debug + PartialEq,
 {
-	fn get_with_key(&self, key: &K) -> Option<&T> {
-		self.iter().find(|(k, _)| k == key).map(|(_, v)| v)
+	pub fn new() -> Self {
+		LimitedFifoMap { limit: DEFAULT_MAX_ITEMS_IN_QUEUE, queue: VecDeque::new() }
+	}
+
+	pub fn with_limit(mut self, limit: usize) -> Self {
+		// cannot set a number smaller than the default.
+		if limit < DEFAULT_MAX_ITEMS_IN_QUEUE {
+			self.limit = DEFAULT_MAX_ITEMS_IN_QUEUE
```

Comment:
Nice catch. Maybe we should also mention why. We can add a comment that says
```
// because if the limit is too small, restoring SPC messages from an archive might have issues since the archived data is inserted in large batches.
```